### PR TITLE
Navigation ASL

### PR DESCRIPTION
**Pull request by pgavigan**
Work in progress of the new ASL program for driving the robot and delivering mail. It is incomplete so far, but incorporates a number of features.

Major piece that is missing: the plans for navigating the specific map (how to decide to turn left, turn right, turn around, go forward, etc).

Work that is implemented so far:
- high level plan for delivering mail, which triggers a sub plan for travelling to a destination
- Line following (which was already there)
- docking, which likely needs to be updated.

This will need some testing and clean up.

I'm very much interested in perspective from @blacktorch on what needs to be done to bring this to a fully working implementation.

### CODE REVIEW

File: asl/demo.asl [hunk 1]
```diff
@@ -1,107 +1,181 @@
...
- +!dock
-    : atDockPost & not moving
-    <- dock_bot.
 
- +!dock.
+/** 
+ * followPath
+ * Follow the line.
+ */
++!followPath
```
Comment (esfandia) [inline asl/demo.asl:237]: is there a way to generalize these rules using unification? Something like:
`+!followPath : line(X) <- drive(X); !followPath.`

**Review by esfandia — COMMENTED** [review]
I think what @blacktorch  is asking is to provide the hard-coded plans specifically for the "4" configuration. So plans that use post1, post2, etc. to tell the agent where to go next.

File: asl/demo.asl (hunk 1 above)
Comment (pgavigan) [reply asl/demo.asl:237]: I like this, Will implement.

**Comment by pgavigan** [comment]
> I think what @blacktorch is asking is to provide the hard-coded plans specifically for the "4" configuration. So plans that use post1, post2, etc. to tell the agent where to go next.

I know, but I wanted to at least get a preliminary wip out there with the skeleton of how this would all fit together. At the time I generated this PR, it was unclear how much more time I would have to work on this today.

**Comment by pgavigan** [comment]
> > I think what @blacktorch is asking is to provide the hard-coded plans specifically for the "4" configuration. So plans that use post1, post2, etc. to tell the agent where to go next.
> 
> I know, but I wanted to at least get a preliminary wip out there with the skeleton of how this would all fit together. At the time I generated this PR, it was unclear how much more time I would have to work on this today.

Map specific rules have been added for determining if the destination is ahead, left, right, or behind. These have not been tested. 

These rules have been notionally added to the plan context for !goToLocation, although this will need to be updated through testing with the robot. Likely using commands that are not implemented.

**Pull request merged**